File: android/app/src/test/java/com/pumplo/app/WatchActionCodecTest.java
package com.pumplo.app;

import static org.junit.Assert.assertEquals;
import static org.junit.Assert.assertNull;

import org.junit.Test;

public class WatchActionCodecTest {

    @Test
    public void decodesLogSetWithWeight() {
        WatchAction a = WatchActionCodec.decode("logSet|40.5|10");
        assertEquals("logSet", a.type);
        assertEquals(Double.valueOf(40.5), a.weight);
        assertEquals(Integer.valueOf(10), a.reps);
    }

    @Test
    public void decodesLogSetWithoutWeight() {
        WatchAction a = WatchActionCodec.decode("logSet|-|8");
        assertEquals("logSet", a.type);
        assertNull(a.weight);
        assertEquals(Integer.valueOf(8), a.reps);
    }

    @Test
    public void decodesSimpleActions() {
        assertEquals("goPrevSet", WatchActionCodec.decode("goPrevSet").type);
        assertEquals("goNextSet", WatchActionCodec.decode("goNextSet").type);
        assertEquals("skipRest", WatchActionCodec.decode("skipRest").type);
        assertEquals("addRest15", WatchActionCodec.decode("addRest15").type);
    }

    @Test
    public void rejectsGarbage() {
        assertNull(WatchActionCodec.decode(""));
        assertNull(WatchActionCodec.decode(null));
        assertNull(WatchActionCodec.decode("logSet|40.5"));
        assertNull(WatchActionCodec.decode("logSet|x|y"));
        assertNull(WatchActionCodec.decode("selfDestruct"));
    }
}

File: android/wear/src/test/java/com/pumplo/wear/WatchActionCodecTest.kt
package com.pumplo.wear

import org.junit.Assert.assertEquals
import org.junit.Test

class WatchActionCodecTest {

    // These literals are the wire contract shared with
    // android/app/src/main/java/com/pumplo/app/WatchActionCodec.java — keep in sync.
    @Test
    fun encodesLogSetWithWeight() {
        assertEquals("logSet|40.5|10", encodeWatchAction(WatchAction.LogSet(40.5, 10)))
    }

    @Test
    fun encodesWholeWeightWithOneDecimal() {
        assertEquals("logSet|40.0|10", encodeWatchAction(WatchAction.LogSet(40.0, 10)))
    }

    @Test
    fun encodesLogSetWithoutWeight() {
        assertEquals("logSet|-|8", encodeWatchAction(WatchAction.LogSet(null, 8)))
    }

    @Test
    fun encodesSimpleActions() {
        assertEquals("goPrevSet", encodeWatchAction(WatchAction.GoPrevSet))
        assertEquals("goNextSet", encodeWatchAction(WatchAction.GoNextSet))
        assertEquals("skipRest", encodeWatchAction(WatchAction.SkipRest))
        assertEquals("addRest15", encodeWatchAction(WatchAction.AddRest15))
    }
}
